flo-runner/src/main/java/com/spotify/flo/context/ForkingEvalContext.java

```diff
@@ -0,0 +1,319 @@
+          throw new RuntimeException(e);
+        }
+
+        final Path closureFile = tempdir.resolve("closure");
+        final Path resultFile = tempdir.resolve("result");
+        final Path errorFile = tempdir.resolve("error");
+
+        final String home = System.getProperty("java.home");
+        final String classPath = System.getProperty("java.class.path");
+        final Path java = Paths.get(home, "bin", "java").toAbsolutePath().normalize();
+
+        try {
+          PersistingContext.serialize(value, closureFile);
+        } catch (Exception e) {
+          throw new RuntimeException("Failed to serialize closure", e);
+        }
+
+        final ProcessBuilder processBuilder = new ProcessBuilder()
+            .command(
+                java.toString(),
+                "-cp", classPath,
+                Trampoline.class.getName(),
+                closureFile.toString(),
+                resultFile.toString(),
+                errorFile.toString())
+            .directory(workdir.toFile());
+
+        final Process process;
+        try {
+          process = processBuilder.start();
+        } catch (IOException e) {
+          throw new RuntimeException(e);
+        }
+
+        // TODO: log output
+        // TODO: propagate grpc context
+        executor.submit(() -> copy(process.getInputStream(), System.out));
+        executor.submit(() -> copy(process.getErrorStream(), System.err));
+
+        final boolean exited;
+        try {
+          // TODO: configurable timeout
+          exited = process.waitFor(1, TimeUnit.DAYS);
+        } catch (InterruptedException e) {
+          process.destroyForcibly();
+          Thread.currentThread().interrupt();
+          throw new RuntimeException(e);
+        }
+        if (!exited) {
+          process.destroyForcibly();
+          throw new RuntimeException("Subprocess timed out");
+        }
+
+        if (process.exitValue() != 0) {
+          throw new RuntimeException("Subprocess failed: " + process.exitValue());
+        }
+
+        if (Files.exists(errorFile)) {
+          // Failed
+          final Throwable error;
+          try {
+            error = PersistingContext.deserialize(errorFile);
+          } catch (Exception e) {
+            throw new RuntimeException("Failed to deserialize error", e);
+          }
+          if (error instanceof RuntimeException) {
+            throw (RuntimeException) error;
+          } else {
+            throw new RuntimeException(error);
+          }
+        } else {
+          // Success
+          final T result;
+          try {
+            result = PersistingContext.deserialize(resultFile);
+          } catch (Exception e) {
+            throw new RuntimeException("Failed to deserialize result", e);
+          }
+          return result;
+        }
+      } finally {
+        executor.shutdownNow();
+        tryDeleteDir(tempdir);
+      }
+    };
+  }
+
+  private void tryDeleteDir(Path path) {
```

Comment:
amazing still so much code needed to recursively delete a dir in Java.

Reply:
My favorite is shelling out to `rm -rf`. Battle tested and hard to beat!